### DISCUSSION TITLE
PostgreSQL with TimescaleDB extension

### DISCUSSION THREAD

**Discussion by xmichaelx** [directus/directus#12167] [Archive]
How can I have access to TimescaleDB hypertable in Directus? Has anyone tried to do that?

```sql
CREATE TABLE IF NOT EXISTS weather_data (
    dt TIMESTAMPTZ NOT NULL,
);
SELECT create_hypertable('weather_data ','dt');
```

Directus requires to have table with primary id which conflicts with `create_hypertable` when I drop primary key constraint from table created in Directus UI it disappears from the UI.

Is there a way to do that in Directus?

**Comment by rijkvanzanten** [comment]
I've never used TimescaleDB before, so no idea if this makes sense, but how does having a primary key conflict with hypertables?

**Reply by martin-braun** [reply]
@rijkvanzanten Could we actually un-archive this discussion? Please let's talk about getting TimescaleDB support to Directus. We are launching a hardware project that runs with Directus software in the backend. Each device will log about 40k entries per month to the database and the initial production batch will be 100 devices, followed by a larger batch of 900 devices.

Our target is to sell all devices until end of next year. At this stage, we already know that our current backend will collapse with an estimate of 40m write operations per month. Our challenge thus is to scale up and TimescaleDB is one possible strategy to do so.

**Reply by rijkvanzanten** [reply]
@martin-braun I'd rather create a new feature request so it follows the correct format 🙂

**Reply by beasteers** [reply]
I'll let @martin-braun dictate their full feature request, but personally the biggest benefit to our team would be being able to manage the table schemas through Directus's UI. 

Right now we are using both Directus's schema export followed by Hasura Migrations (SQL up down scripts for TimescaleDB and other things like Postgres functions) to manage our tables since we are already using Hasura for the API.

Regarding primary keys, we don't set a primary key. Instead we use two unique indexes on two columns:
 * `time, device_id`
 * `time, deployment_id` 
which allows us to do upserts.

It can actually be a problem if someone accidentally starts tracking a hypertable with Directus (by clicking on the crossed-out eye) because then directus will create an empty table when we bootstrap the schema and our Hasura migrations will fail because the table already exists (and has the wrong schema).

**Reply by martin-braun** [reply]
@beasteers You are much deeper into TimescaleDB than I am, I literally learned about it 2 days ago. At this stage I lack the experience with TimescaleDB to outline what I really want or expect from Directus, but you seem to be very concise about what is really missing.

Would you mind forming a feature request for us, or could I just quote you? If you like me to quote you, can you give me anything else that I can add to the feature request that would help the team to understand the requirements, challenges and benefits that come with TimescaleDB?

**Reply by beasteers** [reply]
I'd have to find some time to draft the full RFC, but there's some open questions about the best way forward:

**Should hypertables be treated as just another Collection? In theory it could be as simple as a "Is TimescaleDB Hypertable" checkbox with a few other settings.**
 * hypertables probably don't need a UI though perhaps they should have one for consistency's sake
 * Directus's collection list view may not be equipped to handle the size of hypertables
 * TimescaleDB is only for postgres which unfortunately feels like a losing argument for directus
 * There's the whole issue with primary key vs multi-column unique constraints

**Could it be done through an extension instead?**
 * I've considered a module extension with a significantly simpler interface
 * This would require being able to extend schema import/export functionality from within an extension (think wordpress-style `cleanedSchema = add_filter("schema.import, customSchema)`)

For example, Hasura has this interface:
<img width="1397" alt="image" src="https://github.com/user-attachments/assets/d703bc1e-4f1d-475d-bf1e-d269d4faff28">

Although as I work with these tools more, I'm starting to think that perhaps Directus will always need something like Hasura to handle more DB-specific functionality because there will always be things that can only be built by writing your own queries. 

Hasura's computed fields are an example - they are just postgres functions that you can insert into your graphql schema:

<img width="708" alt="image" src="https://github.com/user-attachments/assets/e2532b6d-70b7-42b6-afb8-4efe30420a6d">

And breaking schema migrations (such as table or column renames) are non-trivial to do with directus currently, fortunately it's not so bad with sequential sql migrations.

This also isn't getting into things like configuring TSDB's table compression, continuous aggregates (i.e. materialized views), anything else I'm missing @jonatas?

--------

This is all to say, I really love Directus+TimescaleDB together and think they combine into a great IoT network backend ❤️ and I'm really hoping the two teams can foster a relationship between each other.

I would love to manage all of our schema with one tool, but the realistic part of me thinks that because of some of Directus's design decisions and some practical realities, we will still need Hasura in the mix to fill in the gaps. I realize Directus is in the process of addressing some of these (renames, node.js-based computed fields, some schema migration updates(?)), but I'm not sure how much of the gap it'll cover.

Curious if @rijkvanzanten has opinions on this.

**Reply by rijkvanzanten** [reply]
The benefit of requiring the user to provide the SQL is that you give an escape hatch for this type of thing, but simultaneously it comes with the major downside that you're now locked into a specific database / version. That in turn complicates things like generated config as code or extensions that are expected to work cross database.

**Reply by beasteers** [reply]
> The benefit of requiring the user to provide the SQL is that you give an escape hatch for this type of thing

I agree with not forcing to write raw queries, but having escape hatches might lighten the load of feature requests.

> but simultaneously it comes with the major downside that you're now locked into a specific database / version.

I don't really see this as a problem from a team perspective because you typically commit to a database. If the migration cost for Directus is to just rewrite a couple of raw queries you wrote, that seems like a reasonable ask. 

From a platform perspective, I understand that you don't want to have marketplace "mongodb extensions" and "postgres extensions". At the same time, different databases have their own features (e.g. timescale 🙂)

---

@rijkvanzanten out of curiosity, if you were to recommend someone to build this as an extension, how would you recommend it be done?

(capitulating to the need for a primary key for directus collections) this is essentially what needs to get run:
```sql
CREATE TABLE IF NOT EXISTS weather_data (
    id SERIAL PRIMARY KEY,  -- auto-incrementing primary key
    time TIMESTAMPTZ NOT NULL,
    device_id TEXT,
    precip_5m_mm INTEGER
);

-- convert from normal table to hypertable
SELECT create_hypertable('weather_data ','time');

-- bonus, but would need a way to configure it (and a place to store the meta)
CREATE UNIQUE INDEX ON weather_data (device_id, time DESC);
```

After a bit of thought, I actually think you could do it by: 
1. creating a Hypertable field ([alias](https://docs.directus.io/user-guide/overview/glossary.html#types) field, like [groups](https://github.com/directus/directus/blob/62f57308f3ea6362866d9da06575c583d7739460/app/src/interfaces/group-raw/index.ts)/[presentation](https://github.com/directus/directus/blob/62f57308f3ea6362866d9da06575c583d7739460/app/src/interfaces/presentation-divider/index.ts)).
2. use field options (or group nesting? can you limit child count?) to configure the 'time' column
3. use `directus_fields.create` ([`<system-collection>.create`](https://docs.directus.io/extensions/hooks.html#available-events)) to run `SELECT create_hypertable`

In theory, a multi-column unique index could be defined as a "field" as well, but I'm curious if this would be seen as a misuse of fields. I hadn't really thought of this use of fields before, but it seems like it opens up a lot of possibilities.

**Reply by jonatas** [reply]
I was thinking about a "timeseries-collection" as a thing that other databases could also implement and make smart decisions like how Timescaledb makes the partition easier for postgres. 

> This also isn't getting into things like configuring TSDB's table compression, continuous aggregates (i.e. materialized views), anything else I'm missing @jonatas?

We have policies that are for refresh aggregates, compression, retention, and tiered storage which moves them to s3. Another "detached" stuff would be able to see the chunks from the hypertable and get some statistics about.

Also, in the `timescaledb_information` schema, there are the `job_stats` and `job_errors` could give some insights and alert about some inconsistency.

**Comment by mslonina** [comment]
Hi,

as far as I understand @xmichaelx, hypertable needs the table to have time-like primary key - this is currently not possible in Directus to set such. On the other hand, you can create a collection, with fields and manually replace primary keys, i.e. setting composite PRIMARY KEY(id, time), but then collection editing capabilities is lost (not displaying any previously configured fields). The other way round - if one create a table without primary key, the Directus interface doesn't detect the columns.

@rijkvanzanten is primary key a requirement for the Directus to detect/manage db tables? One option here seems to be to add support for timestamp-like field as primary key - could you point where to possibly start within the code?

**Reply by rijkvanzanten** [reply]
As long as the table in the database has a column marked primary key, Directus should start using that as the correct primary key in other operations. There's currently no way to select a datetime field as the primary key type from the Directus side, but it should work if you create the table directly in the database first, and have Directus introspect it

**Reply by xmichaelx** [reply]
@rijkvanzanten  Could this be composite primary key build from serial id and timestamp?

**Reply by rijkvanzanten** [reply]
Directus doesn't currently support composite primary keys, so I don't think that'll work as expected

**Reply by jonatas** [reply]
Maybe through UUIDv7?
https://x.com/jamessewell/status/1730125437903450129

Here's also a small ruby example showing how to build a composite id to update: https://ideia.me/using-the-timescale-gem-with-ruby

**Reply by beasteers** [reply]
Looking into it more and it's just currently not possible for that reason. 
 * Directus's Data Model ignores tables without a non-composite primary key
 * TimescaleDB only supports primary keys that include the partitioning key (time)
 * Therefore the only way it works is for the time column to be the sole primary key. But this is a non-starter in most cases. It's the same issue with UUIDv7 because the timestamps you're inserting don't have nanosecond precision and can still collide (e.g. two dates at midnight).

A possible solution is for the Data Model page to [allow](https://github.com/directus/directus/blob/46611e67512279127216ddeec99a810cfb450dce/api/src/utils/get-schema.ts#L137) configuring Collection Schemas regardless of their primary key status (just don't filter them out of the data model). The rest of the app can still ignore collections without a primary key, but the schema editor as it is just breaks as soon as the primary key is no longer singular. Though it sounds like the conversation is waiting on Project Durus.
